File: dicom_to_jpg.js
const fs = require('fs');
const path = require('path');
const dicomParser = require('dicom-parser');
const { createCanvas } = require('canvas');

// Function to convert DICOM files to JPG
async function convertDicomToJpg(dicomPath, jpgPath) {
    try {
        // Read the DICOM file
        const dicomBuffer = fs.readFileSync(dicomPath);
        const dataSet = dicomParser.parseDicom(dicomBuffer);

        // Get pixel data from DICOM
        const pixelDataElement = dataSet.elements.x7fe00010;
        const pixelData = new Uint8Array(dataSet.byteArray.buffer, pixelDataElement.dataOffset, pixelDataElement.length);

        // Create canvas and context
        const canvas = createCanvas(dataSet.uint16('x00280011'), dataSet.uint16('x00280010'));
        const context = canvas.getContext('2d');

        // Create ImageData from pixel data
        const imageData = new ImageData(new Uint8ClampedArray(pixelData), canvas.width, canvas.height);

        // Put ImageData onto canvas
        context.putImageData(imageData, 0, 0);

        // Convert canvas to JPG
        const jpgStream = fs.createWriteStream(jpgPath);
        const jpgStreamData = canvas.createJPEGStream();
        jpgStreamData.pipe(jpgStream);

        console.log(`Converted ${dicomPath} to ${jpgPath}`);
    } catch (error) {
        console.error(`Failed to convert ${dicomPath}: ${error}`);
    }
}

// Specify the paths
const mainFolder = '/home/vishva/Downloads/ANONYMOUS';
const outputMainFolder = '/home/vishva/Downloads/jpg_images';

// Create the output main directory if it doesn't exist
fs.mkdirSync(outputMainFolder, { recursive: true });
console.log(`Created output directory: ${outputMainFolder}`);

// Traverse the directory structure
fs.readdirSync(mainFolder, { withFileTypes: true }).forEach(item => {
    const itemPath = path.join(mainFolder, item.name);
    if (item.isDirectory()) {
        const outputSubFolder = path.join(outputMainFolder, item.name);
        fs.mkdirSync(outputSubFolder, { recursive: true });

        // Traverse subfolder
        fs.readdirSync(itemPath, { withFileTypes: true }).forEach(subItem => {
            const subItemPath = path.join(itemPath, subItem.name);
            if (subItem.isDirectory()) {
                const outputSubSubFolder = path.join(outputSubFolder, subItem.name);
                fs.mkdirSync(outputSubSubFolder, { recursive: true });

                // Convert DICOM files
                fs.readdirSync(subItemPath).forEach(file => {
                    if (file.toLowerCase().endsWith('.dcm')) {
                        const dicomPath = path.join(subItemPath, file);
                        const jpgFile = file.replace('.dcm', '.jpg');
                        const jpgPath = path.join(outputSubSubFolder, jpgFile);

                        convertDicomToJpg(dicomPath, jpgPath);
                    }
                });
            }
        });
    }
});

console.log("All DICOM files have been converted to JPG.");

File: dicom_to_jpg.py
import os
import pydicom
from PIL import Image

# Function to convert DICOM files to JPG
def convert_dicom_to_jpg(dicom_path, jpg_path):
    try:
        print(f"dicom_path: {dicom_path}")
        print(f"jpg_path: {jpg_path}")
        # Read the DICOM file
        dicom_data = pydicom.dcmread(dicom_path)
        
        # Get the pixel array from the DICOM file
        pixel_array = dicom_data.pixel_array
        
        # Convert the pixel array to an image
        image = Image.fromarray(pixel_array)
        
        # Convert the image to RGB mode if it's not already
        if image.mode != 'RGB':
            image = image.convert('RGB')
        
        # Save the image as a JPG file
        image.save(jpg_path)
        print(f"Converted {dicom_path} to {jpg_path}")
    except Exception as e:
        print(f"Failed to convert {dicom_path}: {e}")

# Specify the paths
main_folder = '/home/vishva/Downloads/ANONYMOUS'  
output_main_folder = '/home/vishva/Downloads/jpg_images'

# Create the output main directory if it doesn't exist
os.makedirs(output_main_folder, exist_ok=True)
print(f"Created output directory: {output_main_folder}")

# Traverse the directory structure
for root, dirs, files in os.walk(main_folder):
    print(f"Traversing directory: {root}")
    for file in files:
        # print(f"Found file: {file}")
        if file.lower().endswith('.dcm'):
            dicom_path = os.path.join(root, file)
            # print(f"Found DICOM file: {dicom_path}")
            
            # Construct the corresponding output path
            relative_path = os.path.relpath(dicom_path, main_folder)
            jpg_folder = os.path.join(output_main_folder, os.path.dirname(relative_path))
            os.makedirs(jpg_folder, exist_ok=True)
            print(f"jpg_folder: {jpg_folder}")
            jpg_file = os.path.splitext(os.path.basename(dicom_path))[0] + '.jpg'
            jpg_path = os.path.join(jpg_folder, jpg_file)
            
            # Convert the DICOM file to JPG
            convert_dicom_to_jpg(dicom_path, jpg_path)

print("All DICOM files have been converted to JPG.")
